File: go/nucleotide_count.go
package dna

import "errors"

type Histogram map[rune]int

type DNA string

func (d DNA) Counts() (Histogram, error) {
	h := Histogram{'A': 0, 'C': 0, 'G': 0, 'T': 0}
	for _, c := range d {
		if _, ok := h[c]; ok {
			h[c] += 1
		} else {
			return nil, errors.New("invalid")
		}
	}
	return h, nil
}

File: zig/nucleotide_count.zig
pub const NucleotideError = error{Invalid};

pub const Counts = struct {
    a: u32 = 0,
    c: u32 = 0,
    g: u32 = 0,
    t: u32 = 0,
};

pub fn countNucleotides(s: []const u8) NucleotideError!Counts {
    var ct = Counts{};
    for (s) |c| {
        switch (c) {
            'A' => ct.a += 1,
            'C' => ct.c += 1,
            'G' => ct.g += 1,
            'T' => ct.t += 1,
            else => return NucleotideError.Invalid,
        }
    }
    return ct;
}
